File: src/lib/engine/optimizer.ts
// CNC Cut Plan Engine — Main Optimizer V6

import { TreeNode, Piece } from './types';
import { createRoot, calcPlacedArea, physicalCount, physicalMeasureSet, validatePlacementCandidate, largestFreeRect, cloneTree, consolidateColumns } from './tree-utils';
import { normalizeTree } from './normalization';
import { runPlacement } from './placement';
import { postOptimizeRegroup } from './post-processing';
import {
  groupPiecesBySameWidth,
  groupPiecesBySameHeight,
  groupPiecesByHeight,
  groupPiecesByWidth,
  groupPiecesFillRow,
  groupPiecesFillCol,
  groupPiecesColumnWidth,
  groupPiecesColumnHeight,
  groupPiecesBandFirst,
  groupPiecesBandLast,
  groupByCommonDimension,
  groupByCommonDimensionTransposed,
  groupStripPackingDP,
  groupStripPackingDPTransposed,
  groupCommonDimensionDP,
  groupIdenticalPieces2D,
} from './grouping';

export function getSortStrategies(): ((a: Piece, b: Piece) => number)[] {
  return [
    (a, b) => b.area - a.area || Math.max(b.w, b.h) - Math.max(a.w, a.h),
    (a, b) => Math.max(b.w, b.h) - Math.max(a.w, a.h) || b.area - a.area,
    (a, b) => b.h - a.h || b.w - a.w,
    (a, b) => b.w - a.w || b.h - a.h,
    (a, b) => b.w + b.h - (a.w + a.h),
    (a, b) => b.w / b.h - a.w / a.h,
    (a, b) => Math.min(b.w, b.h) - Math.min(a.w, a.h),
    (a, b) => {
      const ra = Math.max(a.w, a.h) / Math.min(a.w, a.h);
      const rb = Math.max(b.w, b.h) / Math.min(b.w, b.h);
      return rb - ra;
    },
    (a, b) => b.area - a.area || b.w - a.w,
    (a, b) => b.area - a.area || b.h - a.h,
    (a, b) => Math.max(b.w, b.h) - Math.max(a.w, a.h),
    (a, b) => (b.w * b.h) / (b.w + b.h) - (a.w * a.h) / (a.w + a.h),
    // idx 12 — H1: altura ascendente (peça de menor altura primeiro), desempate largura asc.
    // "Menor dimensão primeiro" — arranjo ausente no conjunto (todos os anteriores são desc).
    (a, b) => a.h - b.h || a.w - b.w,
    // idx 13 — H2: largura ascendente (simétrico de H1 no eixo horizontal), desempate altura asc.
    (a, b) => a.w - b.w || a.h - b.h,
  ];
}

export function optimizeV6(
  pieces: Piece[],
  usableW: number,
  usableH: number,
  minBreak: number = 0,
  useGrouping?: boolean,
): { tree: TreeNode; remaining: Piece[] } {
  if (pieces.length === 0) return { tree: createRoot(usableW, usableH), remaining: [] };

  const pieceSignature = (p: Piece) => [
    p.w,
    p.h,
    p.count || 1,
    p.label || '',
    p.groupedAxis || '',
    p.labels?.join(',') || '',
    p.individualDims?.join(',') || '',
  ].join(':');

  const sequenceSignature = (arr: Piece[]) => arr.map(pieceSignature).join('|');

  const strategies = getSortStrategies();

  const rotatedPieces = pieces.map((p) => ({ w: p.h, h: p.w, area: p.area, count: p.count, label: p.label }));

  // Spec 012: peças rotuladas caíam num ramo SEM agrupamento (guard `hasLabels`),
  // perdendo as 50+ variantes abaixo. Como todo trabalho real vem rotulado do
  // relatório de OF, o motor nunca rodava com agrupamento em produção — violando
  // o Princípio III e fragmentando as sobras. O guard encobria duas falhas de
  // conservação na expansão de grupos rotulados, ambas corrigidas: o roteamento
  // do splitAxis (placement.ts) e a mistura de alturas por tolerância no
  // groupStripPackingDP (grouping.ts).
  //
  // Spec 012 (T036): o gate `skipExpensiveGrouping` (pulava as variantes de
  // agrupamento para n>50 e maxRepetition<3) foi REMOVIDO. Ele existia SÓ aqui,
  // sem espelho no Rust — divergência viva do Princípio VI. Como o app roda WASM
  // (que sempre agrupa), o gate só afetava o fallback TS, e nos dados reais do
  // usuário nunca disparava. Removê-lo faz o TS convergir para o WASM: mesma
  // qualidade nos dois motores, ao custo de CPU no fallback TS em jobs grandes de
  // baixa repetição (regime que não ocorre nos relatórios de OF reais).
  const pieceVariantBuilders: Array<() => Piece[]> =
    useGrouping === false
      ? [() => pieces, () => rotatedPieces]
      : [
          // Bug adormecido corrigido (spec 007): estas entradas eram arrays já
          // computados num Array<() => Piece[]>, e buildVariant() lançava
          // TypeError para qualquer chamada sem labels com agrupamento ligado.
          () => pieces,
          () => rotatedPieces,
          () => groupPiecesBySameWidth(pieces, usableH),
          () => groupPiecesBySameWidth(rotatedPieces, usableH),
          () => groupPiecesBySameWidth(pieces),
          () => groupPiecesBySameWidth(rotatedPieces),
          () => groupPiecesBySameHeight(pieces, usableW),
          () => groupPiecesBySameHeight(rotatedPieces, usableW),
          () => groupPiecesBySameHeight(pieces),
          () => groupPiecesBySameHeight(rotatedPieces),
          () => groupPiecesFillRow(pieces, usableW),
          () => groupPiecesFillRow(rotatedPieces, usableW),
          () => groupPiecesFillRow(pieces, usableW, true),
          () => groupPiecesFillRow(rotatedPieces, usableW, true),
          () => groupPiecesFillCol(pieces, usableH),
          () => groupPiecesFillCol(rotatedPieces, usableH),
          () => groupPiecesFillCol(pieces, usableH, true),
          () => groupPiecesFillCol(rotatedPieces, usableH, true),
          () => groupPiecesFillRow(groupPiecesBySameWidth(pieces, usableH), usableW),
          () => groupPiecesFillRow(groupPiecesBySameHeight(pieces, usableW), usableW),
          () => groupPiecesColumnWidth(pieces, usableW),
          () => groupPiecesColumnWidth(rotatedPieces, usableW),
          () => groupPiecesColumnHeight(pieces, usableH),
          () => groupPiecesColumnHeight(rotatedPieces, usableH),
          () => groupPiecesBandFirst(pieces, usableW),
          () => groupPiecesBandFirst(rotatedPieces, usableW),
          () => groupPiecesBandFirst(pieces, usableW, true),
          () => groupPiecesBandFirst(rotatedPieces, usableW, true),
          () => groupPiecesBandLast(pieces, usableW),
          () => groupPiecesBandLast(rotatedPieces, usableW),
          () => groupByCommonDimension(pieces, usableW, usableH),
          () => groupByCommonDimension(rotatedPieces, usableW, usableH),
          () => groupByCommonDimension(pieces, usableW, usableH, 0.3),
          () => groupByCommonDimension(rotatedPieces, usableW, usableH, 0.3),
          () => groupByCommonDimensionTransposed(pieces, usableW, usableH),
          () => groupByCommonDimensionTransposed(rotatedPieces, usableW, usableH),
          () => groupStripPackingDP(pieces, usableW, usableH, 0),
          () => groupStripPackingDP(rotatedPieces, usableW, usableH, 0),
          () => groupStripPackingDP(pieces, usableW, usableH, 5),
          () => groupStripPackingDP(rotatedPieces, usableW, usableH, 5),
          () => groupStripPackingDP(pieces, usableW, usableH, 30),
          () => groupStripPackingDP(rotatedPieces, usableW, usableH, 30),
          () => groupStripPackingDP(pieces, usableW, usableH, 100),
          () => groupStripPackingDP(pieces, usableW, usableH, 5, "raw"),
          () => groupStripPackingDP(rotatedPieces, usableW, usableH, 5, "raw"),
          () => groupStripPackingDPTransposed(pieces, usableW, usableH, 0),
          () => groupStripPackingDPTransposed(rotatedPieces, usableW, usableH, 0),
          () => groupStripPackingDPTransposed(pieces, usableW, usableH, 5),
          () => groupStripPackingDPTransposed(rotatedPieces, usableW, usableH, 5),
          () => groupCommonDimensionDP(pieces, usableW, usableH),
          () => groupCommonDimensionDP(rotatedPieces, usableW, usableH),
          () => groupCommonDimensionDP(pieces, usableW, usableH, 0.2),
          () => groupCommonDimensionDP(rotatedPieces, usableW, usableH, 0.2),
          () => groupIdenticalPieces2D(pieces, usableW, usableH),
          () => groupIdenticalPieces2D(rotatedPieces, usableW, usableH),
        ];

  let bestTree: TreeNode | null = null;
  let bestArea = 0;
  let bestRemaining: Piece[] = [];
  let bestTransposed = false;
  let bestCompactness = Infinity;
  let bestFreeArea = 0; // spec 011 — área do maior retângulo livre do melhor atual
  const seenVariants = new Set<string>();
  const seenSortedOrders = new Set<string>();

  // Spec 012 (T011) — validação no LIMITE candidato→plano. `pieces` é o
  // inventário físico oferecido; toda variante o reagrupa, então a conservação
  // é medida contra este total constante. `validMeasures` guarda as medidas
  // reais para barrar folhas fantasma (INV-2). Um candidato que viole os
  // invariantes é DESCARTADO antes do desempate — não pode vencer por parecer
  // mais compacto (o bug se disfarçando de qualidade, Achado 2).
  const expectedPhysical = physicalCount(pieces);
  const validMeasures = physicalMeasureSet(pieces);
  // Rede de segurança: se NENHUM candidato válido aparecer (patológico — a
  // variante trivial de peças soltas deveria sempre passar), preferimos colocar
  // peças a devolver chapa vazia. Guarda o melhor inválido só para esse caso.
  let fallbackTree: TreeNode | null = null;
  let fallbackArea = 0;
  let fallbackRemaining: Piece[] = [];
  let fallbackTransposed = false;

  /** Fewer top-level columns = more compact waste = better layout */
  function calcCompactness(tree: TreeNode): number {
    const numCols = tree.filhos.length;
    // Penalize layouts with many columns; also prefer fewer total nodes
    let totalNodes = 0;
    function countNodes(n: TreeNode) { totalNodes++; for (const c of n.filhos) countNodes(c); }
    countNodes(tree);
    return numCols * 1000 + totalNodes;
  }

  for (const transposed of [false, true]) {
    const eW = transposed ? usableH : usableW;
    const eH = transposed ? usableW : usableH;

    for (const buildVariant of pieceVariantBuilders) {
      const variant = buildVariant();
      const variantKey = `${transposed ? 'T' : 'N'}|${sequenceSignature(variant)}`;
      if (seenVariants.has(variantKey)) continue;
      seenVariants.add(variantKey);

      for (const sortFn of strategies) {
        const sorted = [...variant].sort(sortFn);
        const sortedKey = `${transposed ? 'T' : 'N'}|${sequenceSignature(sorted)}`;
        if (seenSortedOrders.has(sortedKey)) continue;
        seenSortedOrders.add(sortedKey);

        const result = runPlacement(sorted, eW, eH, minBreak);

        // T011: descartar candidato que viole conservação/fidelidade/rótulo.
        if (!validatePlacementCandidate(result.tree, result.remaining, expectedPhysical, validMeasures)) {
          if (result.area > fallbackArea) {
            fallbackArea = result.area;
            fallbackTree = result.tree;
            fallbackRemaining = result.remaining;
            fallbackTransposed = transposed;
          }
          continue;
        }

        // Área domina (objetivo primário): um candidato com menos área nunca
        // vence, então nem calculamos a consolidação dele (poda de custo).
        if (result.area < bestArea) continue;

        const compactness = calcCompactness(result.tree);

        // Spec 011 — CONSOLIDAÇÃO da sobra (desempate SUBORDINADO à área): entre
        // candidatos de mesma área, preferir o cujo MAIOR retângulo livre é MAIOR
        // — a sobra fica num bloco único reutilizável em vez de fragmentada. A
        // sobra vale por si (definição do usuário), independente de peça específica.
        //
        // CRÍTICO: mede-se na árvore COMO ELA SERÁ finalizada. `normalizeTree`
        // reestrutura os cortes e muda o maior retângulo livre; e o resultado só é
        // normalizado quando transposto (ou minBreak>0). Medir na árvore crua
        // escolheria o candidato errado (medido: âncora ficava em 991k em vez do
        // bloco consolidado). Ver research.md, "Validação empírica".
        //
        // ACHADO que trocou o critério ESCRITO (residual-fit): `runPlacement`
        // DESCARTA a peça que não cabe (`remaining.shift()`), então
        // `result.remaining` é sempre vazio e o residual-fit nunca dispararia; e no
        // cenário-âncora todas as peças cabem (sem "próxima peça") ⇒ só a
        // consolidação distingue os candidatos.
        let measuredTree = result.tree;
        if (transposed) {
          const c = cloneTree(result.tree);
          c.transposed = true;
          measuredTree = normalizeTree(c, usableW, usableH, minBreak);
        } else if (minBreak > 0) {
          measuredTree = normalizeTree(cloneTree(result.tree), usableW, usableH, minBreak);
        }
        const freeRect = largestFreeRect(measuredTree, usableW, usableH);
        const freeArea = freeRect ? freeRect.w * freeRect.h : 0;

        const better =
          result.area > bestArea ||
          (result.area === bestArea && (
            freeArea > bestFreeArea ||
            (freeArea === bestFreeArea && compactness < bestCompactness)
          ));

        if (better) {
          bestArea = result.area;
          bestTree = result.tree;
          bestRemaining = result.remaining;
          bestTransposed = transposed;
          bestCompactness = compactness;
          bestFreeArea = freeArea;
        }
      }
    }
  }

  // T011: só cai no fallback (candidato inválido) se NENHUM válido venceu —
  // colocar peças com um plano imperfeito é melhor que devolver chapa vazia.
  if (!bestTree && fallbackTree) {
    bestTree = fallbackTree;
    bestRemaining = fallbackRemaining;
    bestTransposed = fallbackTransposed;
  }

  let finalTree = bestTree || createRoot(usableW, usableH);
  if (bestTransposed) {
    finalTree.transposed = true;
    finalTree = normalizeTree(finalTree, usableW, usableH, minBreak);
  } else if (minBreak > 0) {
    finalTree = normalizeTree(finalTree, usableW, usableH, minBreak);
  }

  // Spec 013 — "cortar até o final primeiro": consolida a sobra lateral de
  // colunas com peças de mesma largura empilhadas. Não move peças; só junta os
  // retalhos laterais num bloco reutilizável.
  consolidateColumns(finalTree);

  return {
    tree: finalTree,
    remaining: bestRemaining,
  };
}

File: wasm-engine/src/optimizer.rs
use crate::types::{Arena, Piece, OptimizeV6Result, ROOT_ID};
use crate::placement::run_placement;
use crate::normalization::normalize_tree;
use crate::grouping::{
    group_pieces_by_same_width, group_pieces_by_same_height,
    group_pieces_fill_row, group_pieces_fill_col,
    group_pieces_column_width, group_pieces_column_height,
    group_pieces_band_first, group_pieces_band_last,
    group_by_common_dimension, group_by_common_dimension_transposed,
    group_strip_packing_dp, group_strip_packing_dp_transposed,
    group_common_dimension_dp,
};

pub const NUM_SORT_STRATEGIES: usize = 12;

pub fn sort_by_strategy(pieces: &mut Vec<Piece>, idx: usize) {
    pieces.sort_by(|a, b| cmp_by_strategy(a, b, idx));
}

pub fn cmp_by_strategy(a: &Piece, b: &Piece, idx: usize) -> std::cmp::Ordering {
    match idx {
        0 => b.area.partial_cmp(&a.area).unwrap_or(std::cmp::Ordering::Equal)
                .then_with(|| f64::max(b.w, b.h).partial_cmp(&f64::max(a.w, a.h)).unwrap_or(std::cmp::Ordering::Equal)),
        1 => f64::max(b.w, b.h).partial_cmp(&f64::max(a.w, a.h)).unwrap_or(std::cmp::Ordering::Equal)
                .then_with(|| b.area.partial_cmp(&a.area).unwrap_or(std::cmp::Ordering::Equal)),
        2 => b.h.partial_cmp(&a.h).unwrap_or(std::cmp::Ordering::Equal)
                .then_with(|| b.w.partial_cmp(&a.w).unwrap_or(std::cmp::Ordering::Equal)),
        3 => b.w.partial_cmp(&a.w).unwrap_or(std::cmp::Ordering::Equal)
                .then_with(|| b.h.partial_cmp(&a.h).unwrap_or(std::cmp::Ordering::Equal)),
        4 => (b.w + b.h).partial_cmp(&(a.w + a.h)).unwrap_or(std::cmp::Ordering::Equal),
        5 => (b.w / b.h.max(0.001)).partial_cmp(&(a.w / a.h.max(0.001))).unwrap_or(std::cmp::Ordering::Equal),
        6 => f64::min(b.w, b.h).partial_cmp(&f64::min(a.w, a.h)).unwrap_or(std::cmp::Ordering::Equal),
        7 => {
            let ra = f64::max(a.w, a.h) / f64::min(a.w, a.h).max(0.001);
            let rb = f64::max(b.w, b.h) / f64::min(b.w, b.h).max(0.001);
            rb.partial_cmp(&ra).unwrap_or(std::cmp::Ordering::Equal)
        }
        8 => b.area.partial_cmp(&a.area).unwrap_or(std::cmp::Ordering::Equal)
                .then_with(|| b.w.partial_cmp(&a.w).unwrap_or(std::cmp::Ordering::Equal)),
        9 => b.area.partial_cmp(&a.area).unwrap_or(std::cmp::Ordering::Equal)
                .then_with(|| b.h.partial_cmp(&a.h).unwrap_or(std::cmp::Ordering::Equal)),
        10 => f64::max(b.w, b.h).partial_cmp(&f64::max(a.w, a.h)).unwrap_or(std::cmp::Ordering::Equal),
        11 => {
            let va = (a.w * a.h) / (a.w + a.h).max(0.001);
            let vb = (b.w * b.h) / (b.w + b.h).max(0.001);
            vb.partial_cmp(&va).unwrap_or(std::cmp::Ordering::Equal)
        }
        _ => std::cmp::Ordering::Equal,
    }
}

fn calc_compactness(arena: &Arena) -> u32 {
    let num_cols = arena.nodes[ROOT_ID as usize].children.len() as u32;
    let total_nodes = (arena.nodes.len() as u32).saturating_sub(1);
    num_cols * 1000 + total_nodes
}

fn rotate_pieces(pieces: &[Piece]) -> Vec<Piece> {
    pieces.iter().map(|p| Piece {
        w: p.h, h: p.w, area: p.area, count: p.count, label: p.label.clone(),
        labels: p.labels.clone(), grouped_axis: p.grouped_axis.clone(), individual_dims: p.individual_dims.clone(),
    }).collect()
}

pub fn optimize_v6_arena(
    pieces: &[Piece],
    usable_w: f64,
    usable_h: f64,
    min_break: f64,
    use_grouping: bool,
) -> (Arena, Vec<Piece>) {
    if pieces.is_empty() {
        return (Arena::new_root(usable_w), vec![]);
    }

    let has_labels = pieces.iter().any(|p| p.label.is_some());
    let rotated = rotate_pieces(pieces);

    let variants: Vec<Vec<Piece>> = if has_labels || !use_grouping {
        vec![pieces.to_vec(), rotated.clone()]
    } else {
        vec![
            pieces.to_vec(),
            rotated.clone(),
            group_pieces_by_same_width(pieces, usable_h),
            group_pieces_by_same_width(&rotated, usable_h),
            group_pieces_by_same_width(pieces, f64::MAX),
            group_pieces_by_same_width(&rotated, f64::MAX),
            group_pieces_by_same_height(pieces, usable_w),
            group_pieces_by_same_height(&rotated, usable_w),
            group_pieces_by_same_height(pieces, f64::MAX),
            group_pieces_by_same_height(&rotated, f64::MAX),
            group_pieces_fill_row(pieces, usable_w, false),
            group_pieces_fill_row(&rotated, usable_w, false),
            group_pieces_fill_row(pieces, usable_w, true),
            group_pieces_fill_row(&rotated, usable_w, true),
            group_pieces_fill_col(pieces, usable_h, false),
            group_pieces_fill_col(&rotated, usable_h, false),
            group_pieces_fill_col(pieces, usable_h, true),
            group_pieces_fill_col(&rotated, usable_h, true),
            group_pieces_fill_row(&group_pieces_by_same_width(pieces, usable_h), usable_w, false),
            group_pieces_fill_row(&group_pieces_by_same_height(pieces, usable_w), usable_w, false),
            group_pieces_column_width(pieces, usable_w),
            group_pieces_column_width(&rotated, usable_w),
            group_pieces_column_height(pieces, usable_h),
            group_pieces_column_height(&rotated, usable_h),
            group_pieces_band_first(pieces, usable_w, false),
            group_pieces_band_first(&rotated, usable_w, false),
            group_pieces_band_first(pieces, usable_w, true),
            group_pieces_band_first(&rotated, usable_w, true),
            group_pieces_band_last(pieces, usable_w, false),
            group_pieces_band_last(&rotated, usable_w, false),
            group_by_common_dimension(pieces, usable_w, usable_h, 0.4),
            group_by_common_dimension(&rotated, usable_w, usable_h, 0.4),
            group_by_common_dimension(pieces, usable_w, usable_h, 0.3),
            group_by_common_dimension(&rotated, usable_w, usable_h, 0.3),
            group_by_common_dimension_transposed(pieces, usable_w, usable_h, 0.4),
            group_by_common_dimension_transposed(&rotated, usable_w, usable_h, 0.4),
            group_strip_packing_dp(pieces, usable_w, usable_h, 0.0, false),
            group_strip_packing_dp(&rotated, usable_w, usable_h, 0.0, false),
            group_strip_packing_dp(pieces, usable_w, usable_h, 5.0, false),
            group_strip_packing_dp(&rotated, usable_w, usable_h, 5.0, false),
            group_strip_packing_dp(pieces, usable_w, usable_h, 30.0, false),
            group_strip_packing_dp(&rotated, usable_w, usable_h, 30.0, false),
            group_strip_packing_dp(pieces, usable_w, usable_h, 100.0, false),
            group_strip_packing_dp(pieces, usable_w, usable_h, 5.0, true),
            group_strip_packing_dp(&rotated, usable_w, usable_h, 5.0, true),
            group_strip_packing_dp_transposed(pieces, usable_w, usable_h, 0.0),
            group_strip_packing_dp_transposed(&rotated, usable_w, usable_h, 0.0),
            group_strip_packing_dp_transposed(pieces, usable_w, usable_h, 5.0),
            group_strip_packing_dp_transposed(&rotated, usable_w, usable_h, 5.0),
            group_common_dimension_dp(pieces, usable_w, usable_h, 0.3),
            group_common_dimension_dp(&rotated, usable_w, usable_h, 0.3),
            group_common_dimension_dp(pieces, usable_w, usable_h, 0.2),
            group_common_dimension_dp(&rotated, usable_w, usable_h, 0.2),
        ]
    };

    let mut best_arena: Option<Arena> = None;
    let mut best_area = 0.0f64;
    let mut best_remaining: Vec<Piece> = pieces.to_vec();
    let mut best_transposed = false;
    let mut best_compactness = u32::MAX;

    for &transposed in &[false, true] {
        let ew = if transposed { usable_h } else { usable_w };
        let eh = if transposed { usable_w } else { usable_h };

        for variant in &variants {
            for si in 0..NUM_SORT_STRATEGIES {
                let mut sorted = variant.clone();
                sort_by_strategy(&mut sorted, si);
                let result = run_placement(&sorted, ew, eh, min_break, None);
                let compactness = calc_compactness(&result.arena);
                if result.area > best_area
                    || (result.area == best_area && compactness < best_compactness)
                {
                    best_area = result.area;
                    best_remaining = result.remaining;
                    best_transposed = transposed;
                    best_compactness = compactness;
                    best_arena = Some(result.arena);
                }
            }
        }
    }

    let mut final_arena = best_arena.unwrap_or_else(|| Arena::new_root(usable_w));

    if best_transposed {
        final_arena.get_mut(ROOT_ID).transposed = true;
        final_arena = normalize_tree(final_arena, usable_w, usable_h, min_break);
    } else if min_break > 0.0 {
        final_arena = normalize_tree(final_arena, usable_w, usable_h, min_break);
    }

    (final_arena, best_remaining)
}

pub fn optimize_v6(
    pieces: &[Piece],
    usable_w: f64,
    usable_h: f64,
    min_break: f64,
    use_grouping: bool,
) -> OptimizeV6Result {
    let (arena, remaining) = optimize_v6_arena(pieces, usable_w, usable_h, min_break, use_grouping);
    OptimizeV6Result {
        tree: arena.to_json_node(ROOT_ID),
        remaining,
    }
}
